Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp

```diff
@@ -0,0 +1,266 @@
+    eglDestroyImage(eglDisplay, image);
+
+#if USE(GTK4)
+    texture[0] = adoptGRef(gdk_gl_texture_new(context.get(), textureID, size.width(), size.height(), nullptr, nullptr));
+    texture[1] = adoptGRef(gdk_gl_texture_new(context.get(), textureID, size.width(), size.height(), nullptr, nullptr));
+    textureIndex = 0;
+#else
+    textureSize = WTFMove(size);
+#endif
+}
+
+AcceleratedBackingStoreDMABuf::Texture::~Texture()
+{
+    if (textureID) {
+        gdk_gl_context_make_current(context.get());
+        glDeleteTextures(1, &textureID);
+    }
+}
+
+int AcceleratedBackingStoreDMABuf::Texture::width() const
+{
+#if USE(GTK4)
+    return gdk_texture_get_width(texture[0].get());
+#else
+    return textureSize.width();
+#endif
+}
+
+int AcceleratedBackingStoreDMABuf::Texture::height() const
+{
+#if USE(GTK4)
+    return gdk_texture_get_height(texture[0].get());
+#else
+    return textureSize.height();
+#endif
+}
+
+void AcceleratedBackingStoreDMABuf::configure(UnixFileDescriptor&& fd, int fourcc, int32_t offset, int32_t stride, WebCore::IntSize&& size)
+{
+    if (!makeContextCurrent())
+        return;
+
+    m_pendingTexture = makeUnique<Texture>(m_gdkGLContext.get(), WTFMove(fd), fourcc, offset, stride, WTFMove(size));
+}
+
+void AcceleratedBackingStoreDMABuf::frame(CompletionHandler<void()>&& completionHandler)
+{
+    RELEASE_ASSERT(!m_frameCompletionHandler);
+    m_frameCompletionHandler = WTFMove(completionHandler);
+
+    if (m_pendingTexture)
+        m_committedTexture = WTFMove(m_pendingTexture);
+#if USE(GTK4)
+    else
+        m_committedTexture->textureIndex = !m_committedTexture->textureIndex;
+#endif
+    gtk_widget_queue_draw(m_webPage.viewWidget());
+}
+
+void AcceleratedBackingStoreDMABuf::unrealize()
+{
+    if (!m_glContextInitialized)
+        return;
+
+    m_pendingTexture = nullptr;
+    m_committedTexture = nullptr;
+
+    if (m_gdkGLContext && m_gdkGLContext.get() == gdk_gl_context_get_current())
+        gdk_gl_context_clear_current();
+
+    m_glContextInitialized = false;
+}
+
+void AcceleratedBackingStoreDMABuf::tryEnsureGLContext()
+{
+    if (m_glContextInitialized || !gtk_widget_get_realized(m_webPage.viewWidget()))
+        return;
+
+    m_glContextInitialized = true;
+
+    GUniqueOutPtr<GError> error;
+#if USE(GTK4)
+    m_gdkGLContext = adoptGRef(gdk_surface_create_gl_context(gtk_native_get_surface(gtk_widget_get_native(m_webPage.viewWidget())), &error.outPtr()));
+#else
+    m_gdkGLContext = adoptGRef(gdk_window_create_gl_context(gtk_widget_get_window(m_webPage.viewWidget()), &error.outPtr()));
+#endif
+    if (m_gdkGLContext) {
+#if USE(OPENGL_ES)
+        gdk_gl_context_set_use_es(m_gdkGLContext.get(), TRUE);
+#endif
+        gdk_gl_context_realize(m_gdkGLContext.get(), &error.outPtr());
+        if (!error)
+            return;
+    }
+
+    g_warning("GDK is not able to create a GL context, falling back to ... (slow!): %s", error->message);
+    // FIXME: implement fallback?
```

Comment:
Another placeholder discussion just to ensure we don't forget the FIXMEs

Reply:
So, the question here is whether we really want to keep using hardware acceleration when GTK can't. I think we could try to ensure GTK will be able to create a context in checkRequirements() (or even HardwareAccelerationManager) and simply disable hw acceleration if that fails.